### ISSUE TITLE
Pods are not getting migrated to ready nodes if one of the cluster node becomes in "NotReady" Status

### ISSUE THREAD

**Issue by Richa-Srivastava** [pravega/pravega-operator#171]
**Pods are not getting migrated to ready nodes if one of the pks cluster node becomes `NotReady` Status, the pods corresponding to that node became unknown state rather than migrating to other `Ready` nodes:**

Environment details: PKS / K8 with medium cluster:
```
3 master nodes @ large.cpu (4 CPU, 4 GB Ram, 16 GB Disk)
5 worker nodes @ xlarge.cpu(8 cpu, 8 GB Ram, 32 GB Disk)
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
```
root@osboxes:/home/osboxes# kubectl get nodes
NAME                                   STATUS     ROLES     AGE       VERSION
7ea111b5-73b2-4a6a-8992-91eb54bc6f00   Ready      <none>    13d       v1.12.7
83ba7b89-b5a2-4293-b59a-a12267772f83   Ready      <none>    13d       v1.12.7
a2a4a917-a6d6-49cc-b252-ae8a0891448d   Ready      <none>    9d        v1.12.7
d71da967-5c65-48e1-b659-6f709879af5f   Ready      <none>    13d       v1.12.7
d8e8be3a-f191-4929-8bbd-1f530dd777c0   NotReady   <none>    13d       v1.12.7
```
Following is the pods status:
```
# kubectl get po --all-namespaces
NAMESPACE           NAME                                                 READY     STATUS      RESTARTS   AGE
default             pravega-benchmark                                    1/1       Unknown     0          4h
default             pravega-bookie-0                                     1/1       Running     0          3h
default             pravega-bookie-1                                     1/1       Unknown     0          3h
default             pravega-bookie-2                                     1/1       Running     0          3h
default             pravega-operator-779879b48-z9hck                     1/1       Running     0          3h
default             pravega-pravega-controller-84948d8895-j2vhz          1/1       Running     0          3h
default             pravega-pravega-controller-84948d8895-ttvsl          1/1       Running     0          3h
default             pravega-pravega-segmentstore-0                       1/1       Running     0          3h
default             pravega-pravega-segmentstore-1                       1/1       Running     0          2h
default             pravega-pravega-segmentstore-2                       1/1       Running     0          2h
default             pravega-zk-0                                         1/1       Running     0          3h
default             pravega-zk-1                                         1/1       Unknown     0          3h
default             pravega-zk-2                                         1/1       Running     0          3h
default             pravega-zk-3                                         1/1       Running     0          3h
default             shaka-zulu-nfs-client-provisioner-59d7f8f84c-ckkmx   1/1       Running     0          3h
default             zookeeper-operator-685bfcbbc5-4k8t8                  0/1       Running     0          10m
default             zookeeper-operator-685bfcbbc5-kdfgc                  1/1       Unknown     0          3h
kube-system         heapster-85647cf566-nnhzz                            1/1       Running     0          10m
kube-system         heapster-85647cf566-t46rl                            1/1       Unknown     0          13d
kube-system         kube-dns-7559c96fc4-kq5n9                            3/3       Running     0          13d
kube-system         kubernetes-dashboard-5f4b59b97f-hrwmz                1/1       Running     1          9d
kube-system         metrics-server-555d98886f-cp9k6                      1/1       Running     0          13d
kube-system         monitoring-influxdb-cdcf4674-cgxzp                   1/1       Running     0          13d
kube-system         tiller-deploy-95d654d46-dfqh8                        1/1       Running     0          3h
pks-system          cert-generator-v0.11-6hg96                           0/1       Completed   0          13d
pks-system          event-controller-56f9597df8-hq2nq                    2/2       Running     1          13d
pks-system          fluent-bit-6l4hv                                     2/2       Running     0          13d
pks-system          fluent-bit-8hfpj                                     2/2       NodeLost    0          13d
pks-system          fluent-bit-svnb2                                     2/2       Running     0          9d
pks-system          fluent-bit-tdxnj                                     2/2       Running     0          13d
pks-system          fluent-bit-vml52                                     2/2       Running     0          13d
pks-system          sink-controller-65595c498b-dwrhc                     1/1       Running     0          10m
pks-system          sink-controller-65595c498b-k685f                     1/1       Unknown     3          13d
pks-system          telemetry-agent-559f9c8855-55fb8                     1/1       Running     0          21h
pravega-longevity   small-key-manager-7d54fbf9f7-jx8ln                   1/1       Unknown     0          2h
pravega-longevity   small-key-manager-7d54fbf9f7-tw7zn                   1/1       Running     0          10m
pravega-longevity   small-key-worker-57b474f5cb-sd4l9                    1/1       Running     0          10m
pravega-longevity   small-key-worker-57b474f5cb-xgwlk                    1/1       Unknown     0          2h
```

Attaching the complete logs taken during this operation
[pravega-k8s-logging-2019-04-29-16-04-43.zip](https://github.com/pravega/pravega-operator/files/3127247/pravega-k8s-logging-2019-04-29-16-04-43.zip)

**Comment by adrianmo** [comment]
@Richa-Srivastava how long did you wait since the node became not ready? According to [this document](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/), the default pod eviction timeout is 5 minutes.

**Comment by Richa-Srivastava** [comment]
@adrianmo it took more than one hour for pods to show in ready state.

**Comment by Tristan1900** [comment]
I have logged into the master node of my PKS cluster and checked the config in `kube-controller-manager` and I don't see the pod eviction timeout is set, so it should be the default 5min as @adrianmo mentioned. (Not sure if it is ok to put the configuration here)

I also tried `kubectl drain` and `kubectl taint` but the cluster can reschedule the pods to the available nodes. 

It is possible that this problem is related to this active [issue](https://github.com/kubernetes/kubernetes/issues/55713) reported in the Kubernetes community. It seems to be a kubelet bug but there is no fix right now. 

Do you think we should close it or keep it open for tracking that k8s issue @adrianmo ? Thanks!

**Comment by adrianmo** [comment]
@Tristan1900 you are trying to reproduce the issue with user-initiated actions (`drain` and `taint`), which might not be causing the node to become `NotReady`. I think the original issue will not be manifested unless the node becomes unavailable out of the scope of Kubernetes. Can you give it a try?

**Comment by Tristan1900** [comment]
@adrianmo Thanks for the heads up. You are right,  I have tried to reboot a worker node and I found the reasons for the observed behaviors in this issue. 
1.  Stateless deployments actually get migrated
```
pravega-longevity   small-key-manager-7d54fbf9f7-jx8ln                   1/1       Unknown     0          2h
pravega-longevity   small-key-manager-7d54fbf9f7-tw7zn                   1/1       Running     0          10m
pravega-longevity   small-key-worker-57b474f5cb-sd4l9                    1/1       Running     0          10m
pravega-longevity   small-key-worker-57b474f5cb-xgwlk                    1/1       Unknown     0          2h
pks-system          sink-controller-65595c498b-dwrhc                     1/1       Running     0          10m
pks-system          sink-controller-65595c498b-k685f                     1/1       Unknown     3          13d
```
From above, we can see that these stateless pods get migrated.

2. Statefulset do not migrate
```
default             pravega-zk-1                                         1/1       Unknown     0          3h
default             pravega-bookie-1                                     1/1       Unknown     0          3h
```
We know that zk and bk clusters are using statefulset. By design, these pods are not supposed to migrate when the node is lost, which is to avoid multiple instances of pods when the lost node comes back, see [here](https://github.com/kubernetes/kubernetes/issues/54368) and [here](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/pod-safety.md#proposed-changes).

3. Operator pods are in running state but never become ready.
```
default             zookeeper-operator-685bfcbbc5-4k8t8                  0/1       Running     0          10m
default             zookeeper-operator-685bfcbbc5-kdfgc                  1/1       Unknown     0          3
```
What happens is when the operator pod is lost, it gets rescheduled on another node, but there is a leader election locking implementation that prevents the new operator pod from becoming ready. When operator is deployed, there is also a `configmap` called `<opeartor name>-lock` gets created and any new operators will check if that `configmap` exists before actually serving. So in our case, the rescheduled operator will not become ready because the `configmap` is not deleted. See [ref](https://github.com/operator-framework/operator-sdk/blob/121664e0093774c6d4fa4528fe7367d0848e6d74/doc/proposals/leader-for-life.md).

4. Unknown behavior
```
default             pravega-benchmark                                    1/1       Unknown     0          4h
```
I suppose this is a deployment or just a pod. In theory, it should get rescheduled. The possible reason might be the resource is not enough in the cluster since one node is down. This is just an assumption which could not be confirmed by the log.

In summary, I think what we see in this issue is expected. To fix the lost node, the cluster admin could manually delete the node from Apiserver, so that all the pods on that node will be safely deleted and rescheduled later. 
WYDT? @adrianmo

**Comment by Raghuveera-Rao** [comment]
The issue is able to reproduce with VM reset from vcentre UI.
This is k8s medium cluster 
```
3 master nodes @ large.cpu (4 CPU, 4 GB Ram, 16 GB Disk)
5 worker nodes @ xlarge.cpu(8 cpu, 8 GB Ram, 32 GB Disk)
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
Look for the hardware vm from bosh pointing to one of the worker node. 
We will get vm id like 'vm-0163d557-f274-4488-bf7c-72e29671a063' for one worker node (in this case node 4d9a5762-f2e4-4622-a363-8e0612dd0ed4)

(bosh.id=a6078491-e7c5-45d3-9f41-1b18771c4df6)
Reset the VM from vcentre UI.

$ kubectl get nodes
```
NAME                                   STATUS     ROLES     AGE       VERSION
4d9a5762-f2e4-4622-a363-8e0612dd0ed4   NotReady   <none>    13d       v1.12.7
73ee54b3-6ebc-495c-9407-684b0ae79f9e   Ready      <none>    13d       v1.12.7
a9989b53-8219-4cb9-8daf-5587f94dfdec   Ready      <none>    13d       v1.12.7
c8b12cf5-a6f5-4887-83b9-4837c7f0d06c   Ready      <none>    13d       v1.12.7
cf301c7a-a5c8-4ca6-af50-411557233a2a   Ready      <none>    13d       v1.12.7
```
$kubectl get pods
```
pravega-bookie-0                                     1/1       Running   0          1d
pravega-bookie-1                                     1/1       Running   0          1d
pravega-bookie-2                                     1/1       Running   0          1d
pravega-bookie-3                                     1/1       Unknown   0          1d
pravega-operator-5d94f77d4b-jkx2c                    1/1       Running   0          1d
pravega-pravega-controller-6dd4686578-76p7z          1/1       Running   0          1d
pravega-pravega-controller-6dd4686578-gsbdv          1/1       Running   0          1d
pravega-pravega-segmentstore-0                       1/1       Running   0          1d
pravega-pravega-segmentstore-1                       1/1       Running   0          1d
pravega-pravega-segmentstore-2                       1/1       Unknown   0          1d
pravega-zk-0                                         1/1       Running   0          1d
pravega-zk-1                                         1/1       Unknown   0          1d
pravega-zk-2                                         1/1       Running   0          1d
shaka-zulu-nfs-client-provisioner-59d7f8f84c-lmlc2   1/1       Running   0          1d
zookeeper-operator-685bfcbbc5-wx6fc                  1/1       Running   0          1d
```
Node is in **NotReady**  status and pods which are part of this node are in **Unknown** status from more than 15 hours.

**Comment by adrianmo** [comment]
The behavior observed in this node failure scenario corresponds with the expected behavior in Kubernetes ([ref](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/)).  We will elaborate and document the procedure to manually recover a Pravega cluster and operator state with `kubectl` commands.
